### ISSUE TITLE
Centralized place to cancel component removal/route change?

### ISSUE THREAD

**Issue by doesntgolf** [MithrilJS/mithril.js#2113]
## Current Behavior
Right now, the only way to prevent a route change (as I understand it) is to either:

1. use `m.route.set(...)` or `history.back()` or similar from all possible `onmatch`s, or
2. use `preventDefault()` or similar on all possible things that would cause a route change.

(I hope maybe I'm missing something else that makes this easy -- please let me know if this is the case.)

## Possible Solution
My intuition tells me that returning `Promise.reject()` from `onbeforeremove` should prevent the route change (since returning a promise here can already be used to delay removing the component).

I've noticed that when returning a promise in `onbeforeremove`, the URL still changes to the new URL immediately, so I'm guessing other logic is also fired immediately. So if my suggested solution above were to be implemented, I'm guessing this logic would have to be changed to run only after the promise resolves.

## Context
The two ways to prevent route changes that I listed in Current Behavior require attaching the prevention mechanism on all possible `onmatch`s or external links, instead of just on the relevant component itself.

I have a page with a form where I'd like to use something like `window.confirm` to prevent routing away if there are unsaved changes. As it is, I can use `window.confirm` in `onbeforeremove` to ask if the user would like to submit the form or not, but not to cancel the route change entirely if they're not ready yet. 

## Your Environment

* Version used: 1.1.6
* Browser Name and version: Chromium 63
* Operating System and version (desktop or mobile): Fedora 27

**Comment by barneycarroll** [comment]
You sum up the implementation possibilities correctly: either the route endpoint itself or the route navigation mechanism must be used to take this action.

The particular problem you describe where `onmatch` triggers before `onbeforeremove` or `onremove` is a result of the top-down logic of vtree patching: the fact that a vnode in the old tree is removed is a consequence of it being determined not to be present after generating the new tree.

**Comment by doesntgolf** [comment]
> The particular problem you describe where onmatch triggers before onbeforeremove or onremove is a result of the top-down logic of vtree patching: the fact that a vnode in the old tree is removed is a consequence of it being determined not to be present after generating the new tree.

Ahh, that makes a lot of sense, and helps me understand why `onbeforeremove` works the way it does. Thanks!

Is it safe to assume then that my suggestion isn't really feasible?

**Comment by barneycarroll** [comment]
I'm afraid so. Mithril's routing module isn't easy to abstract over. The best I can suggest is an event handler registered by the component that detects attempted navigation and cancels it during the event capture phase, before the event bubbling phase, thus preventing Mithril's routing logic from ever triggering.

[Small working demo here](https://flems.io/#0=N4IgtglgJlA2CmIBcBWAjAOhQFgDQgGd4EBjAF3imRACFYB7EgawgDsBzAYXrAAd7W8VmQwArAiHwAzCAglIA2qFYBDMIiS0GzNlx79BwsRPwkBFYdTOsCZAAR1GLDtz4Ch9gLx3gAHVZ2dgBuEPAA7kh2ABQAlHaeAHx2Cv6BgWBRAOQAFgDMmbh2mY46HHZmboZkmTG4qWn16Vm8BUUAKgBOAJ52qiHsKmS6diphKl0A-HYJSAAiNfUAunUBdvUCbBBkkUHxSX6rgQD0R3Zt2fB2ArA9zlBXUr0q-YMQAnZk2RAEdmGysHZeB14EEPHZvh9soMPh0IOx2PBgVBGnYTnYAEY9EiwCCldhXAJbeBgH5-T52AAG2WBUgpgI69F4iKG8AIGBRKhgAFFQcIADLfCyIqIowKZbG4pgFUWo04AZWelyhrDgiI+9Ds2hUANsg0uBA1YUuJBUAWBYHooPB9lgeo6yTQixlQQwuooGGVqvt3kuiR8MsCECkUXgGDIKg6CJE1PgjwAZHG7ABCawyDoZTIAQWBdi69AArnYCPngRMajEDmkq2k0QA5eCUdX0kFgqKqIZW9HwKEhAsdOKJ2yM6IAWS21NkGLYUF0BBiAbSoaBLeEs1jKnzsDIsQXgVDQ94AAUGbwVAMhgId4dqwBfGU3lbV2VnDr5y7qU0-E28Mgl+CRLsTXzIhIUuHFbCENUIXRAsVSbT5Lj+FV6DCR9qzRPs7E9BB7T+WAAWBA1YE7WN6BzLY7CjH4yA1LV8K6ABaGD8zg8ChQ6AgZTIV94DQwJ51WB96nWVhzUtf9gj2f1rzRBRHRRMTQR5DwBQgwQOiyCVmFaF03VDbDEUKbi3wEwIhNYG9JEIYh4HIN4bGoNAACYkFyABOEAH2UNQNBAAAleh6DIVwDA8YwrOsCwyCsARbDsAKgpC9xhHiaTAhCcJIliKSUQzGCoC6BRbC6BBPCkcwGKkNRZC6JACE-BiiFhKRFmla8mhyNBWkyAL8woSJWjADAGT60Mo1iWp-BlDMWj46tMlrZ44T1dUBpWBcFEyI4usKLanMyRYMDAFReCiGNHkSXc7AzFRWkrJ9q3OuaHoJEhgT1SIhpG90cVYJhntvAGn3Oq7TIemJhsofMSHgKIohUQp0TiS72urBQEaKK57VadEnVR-iZUm1hpqiEpnD0SoPDB-iVjvVgrKIUgLwczQnIABiQJyADZPNwbz1GoE7eHC0xzA8agvoLCgoigRh83UIx8q6Xbtru+otq6yIEuC-RkrINC9syLXAp1ynhFpmJPOWEBfqYeQlBAVQBc0SBPlhWArJLD3NGyMgyF4AgkBOFjeCYdgMAqI5XYnWAAAFMEwLmo-Hd2RZAMguiZahxCtm8gA).

The essential pattern is:
```js
const BlockingComponent = {
  view: () => [
    m('h3', 'Blocking component'),
    
    m('p', 'Try navigating away? >:D')
  ],
  
  oninit: v => {
    // The only kind of navigation this will prevent is that triggered
    // by clicking on items with `href` properties.
    addEventListener(
      'click',
      // Save handler to local state so we can remove it later [1]
      v.state.handler = e => {
        if(e.target.href && !confirm('Are you sure?')){
          // Need to prevent (native behaviour) & stop (Mithril bindings)
          e.preventDefault()
          e.stopPropagation()
        }
      },
      // True means capture: because the listener is bound to the window,
      // our handler will resolve before it gets to locally-bound listeners
      true,
    )
  },
  
  onremove: v => {
    // [1]
    removeEventListener('click', v.state.handler, true)
  },
}
```

**Comment by pygy** [comment]
We really need to add support for more event options at the vdom level.

**Comment by barneycarroll** [comment]
@pygy I don't recall if you were involved in the conversation but @CreaturesInUnitards brought this sentiment up a few days ago in chat (prompted by another neat example of `useCapture`).

My irk with the idea is that I'm not convinced our ingenuity with such a broad and hairy API as the DOM events model can necessarily adequately simplify intent without more things to remember (ie, knowledge both of the subtleties of the underlying native API *and* Mithril's esoteric abstraction).

**Comment by pygy** [comment]
@spacejack had IIRC proposed a nice API to support other options like `passive`.

With the v2 async redraws, I don't think it will be that problematic.

**Comment by barneycarroll** [comment]
I'm closing this issue as a functional duplicate of #1759
